docs/configuration_files.rst
How to Configure Different Aspects of NOEMA_Combine
===================================================

Folders and File Handling
-------------------------

The main configuration options for NOEMA_Combine are set using configuration files. 
These files allow you to specify folders, file handling options, and other 
parameters used throughout the package.
The file name is `config.ini` and an example configuration file is shown below.   

.. code-block:: ini

    [folders]
    uvt_dir = D/
    uvt_dir_out = D30m/
    dir_30m = 30m/
    inputdir = raw_data/, additional_data/, more_data/

    [catalogues]
    line_catalogue = line_catalogue_3mm.csv
    source_catalogue = region_catalogue.yml

    [file_handling]
    ignorefiles_1 = raw_data/FTSOdp20220220.30m
    ignorefiles_2 = raw_data/FTSOdp20220731.30m

    [file_extensions]
    selfcal = _sc
    uvsub = _contsub

The different sections of the configuration file are described below:

- **[folders]**: This section specifies the directories used for input and output data. 
    - `uvt_dir`: Folder with calibrated uv-tables.
    - `uvt_dir_out`: Folder to store calibrated uv-tables output.
    - `dir_30m`: Folder to store calibrated and gridded 30m data.
    - `inputdir`: list of directories on which to search for raw 30m data.
- **[catalogues]**: This section allows you to specify the line and source catalogue files used in the analysis.
    - `line_catalogue`: File containing the line catalogue. It is a CSV file with the list of lines, spectroscopic parameters and processing parameters (see below).
    - `source_catalogue`: File containing the source catalogue. It is a YAML file with the list of sources and their properties (see below).
- **[file_handling]**: Here, you can list files to be ignored during processing. 
  This is useful for excluding specific datasets that may not be relevant, or to avoid unreliable scans.
- **[file_extensions]**: This section defines custom file extensions for self-calibrated and continuum-subtracted files.


Avoid Bad 30m Scans
-------------------

In some cases, certain scans from the 30m telescope may be unreliable or contain bad data.
To avoid processing these bad scans, you can specify them in the configuration file
under the **[file_handling]** section using the ``ignorefiles_1`` and ``ignorefiles_2`` options.
For example:

.. code-block:: ini

    [file_handling]
    ignorefiles_1 = raw_data/FTSOdp20220220.30m


The following ``CLASS`` script will create a new 30m file excluding the bad scans. 
Here from scan 1 until 53, the **LO** and **UO** scans are compromised and 
will be removed.
The output file will be named ``FTSOdp20220220_fix.30m`` and when combined with 
**[file_handling]** this ensures that only the right scans are processed to 
make the cubes.

.. code-block:: fortran

    !
    ! script to comment remove compromised scans seen on 2022-feb-21
    !
    file in FTSOdp20220220.30m
    file out FTSOdp20220220_fix.30m single

    set scan 1 53
    ! only inner units are kept
    set telescope 30ME0*I-*
    find
    get zero
    sic message class s-i
    for i 1 to found
        get n
        write
    next

    !
    ! now save all other data
    !
    set scan 54 *
    set telescope *
    find
    get zero
    for i 1 to found
        get n
        write
    next
    sic message class s+i


Line and Source Catalogues
--------------------------
The line and source catalogues are essential components of the NOEMA_Combine package,
providing necessary information for data processing and analysis.  

- **Line Catalogue**: This is a CSV file that contains a list of spectral lines along with their spectroscopic parameters and processing parameters. 
  Each row in the CSV file corresponds to a different spectral line, and the columns include information such as line frequency, transition details, and any specific processing instructions.
- **Source Catalogue**: This is a YAML file that contains a list of sources and their properties. 
  Each entry in the YAML file represents a different source, with details such as source


Source Catalogue
^^^^^^^^^^^^^^^^

The source catalogue is a YAML file that contains a list of sources and their properties.
Each entry in the YAML file represents a different source, and 
an example of a source catalogue entry in YAML format is shown below:

.. code-block:: yaml

    L1448N:
    RA0: "3h25m36.44s"
    Dec0: "30d45m18.3s"
    height: "33 arcsec"
    width: "30 arcsec"
    fig_width: 6.0
    fig_height: 6.0
    Vlsr: 9.0
    source_30m: "L1448N"
    source_out: "L1448N"
    B5-IRS1:
    RA0: "03h47m41.591s"
    Dec0: "32d51m43.672s"
    height: "40 arcsec"
    width: "40 arcsec"
    fig_width: 6.0
    fig_height: 6.0
    Vlsr: 9.0
    source_30m: "B5-IRS1 B5-Cond*"
    source_out: "B5-IRS1"

- The ``RA0`` and ``Dec0`` fields specify the coordinates of the source to be used for the single dish cubes.
- The ``Vlsr`` is also used for the 30m gridding, and it is in ``km/s``, it is also used to extract uv-tables around the systemic velocity of the source.
- The ``source_30m`` field is used to identify the 30m data files corresponding to the source. It is a string that can include a list of names to be later used in the ``find /source source_30m`` within ``CLASS`` 
- The ``source_out`` field specifies the name to be used for output files related to the source.

Line Catalogue
^^^^^^^^^^^^^^

The line catalogue is a CSV file that contains a list of spectral lines along with their spectroscopic parameters and processing parameters. 
The first line of the CSV file contains the column headers, and each subsequent row corresponds to a different spectral line.

An example of a line catalogue entry in CSV format is shown below:

.. code-block:: text

    #name,QN(filename),freq(GHz),mol(plot),QN(plot),Aul(logs^-1),Eul(K),cat,NOEMAbb,unit,width(km/s),30msetup,30mbb,30mwidth(km/s),30mline(km/s)
    SO,5_5_4_4,215.22065300,SO,"5(5)-4(4)",-3.91491,44.10371,JPL,lo,l009l048,10.0,Setup1,LO,20,5.0
    DCO+,3_2,216.11258220,DCO\u+,"3-2",-2.62283,20.74365,CDMS,lo,l012l051,10.0,Setup2,LO,20,3.0
    c-HCCCH,3_3_0_2_2_1,216.27875600,c-C\d3H\d2,"3(3,0)-2(2,1)",-3.59211,19.46594,CDMS,lo,l013l052,10.0,Setup2,LO,20,3.0


- The ``#name`` column specifies the common name of the spectral line to be used as identifing the molecule and used in the filename.
- The ``QN(filename)`` column indicates the quantum numbers used for the line identification and in the filename.
- The ``freq(GHz)`` column provides the frequency of the line in ``GHz``.
- The ``mol(plot)`` and ``QN(plot)`` columns specify the molecule and quantum numbers to be used for the ``GILDAS`` header.
- The ``Aul(logs^-1)`` and ``Eul(K)`` columns to provide the Einstein A coefficient and upper energy level of the transition, good information to access easily.
- The ``cat`` column indicates the catalogue source (e.g., JPL, CDMS).
- The ``NOEMAbb`` column specifies whether the line is observed with the lower sideband (``lo`` or ``li``) or upper sideband (``ui`` or ``uo``).
- The ``unit`` column indicates the unit for processing. This is the narrow unit name, this is used for the folder and filename.
- The ``width(km/s)`` column specifies the velocity width for processing in ``km/s``. It is used to 
- The ``30msetup`` and ``30mbb`` columns refer to the 30m setup (e.g., Setup1 or Setup2) and the unit that covers the line (``LO``, ``LI``, ``UI``, or ``UO``). This information is not used currently when serching for the scans covering the lines.
- The ``30mwidth(km/s)`` and ``30mline(km/s)`` columns provide parameters specific to 30m data processing. 
    - ``30mwidth(km/s)`` is the with of the cube to be made around the line in ``km/s`` (cubes covers ``Vlsr - 30mwidth(km/s)`` until ``Vlsr + 30mwidth(km/s)``).
    - ``30mline(km/s)`` is the velocity width used to define the window in ``CLASS`` to avoid the line for baseline fitting. The baseline is using the following range of velocities: from ``Vlsr - 30mwidth(km/s)`` until ``Vlsr - 30mline(km/s)``  and ``Vlsr + 30mline(km/s)`` until ``Vlsr + 30mwidth(km/s)``.

config.py
# Configuration file to be used in processing the data.
# This file is used to define the source, the source name, the source coordinates, and the VLSR.
#
source_catalogue = {
    'B5': {'source': 'B5*', 'source_out': 'B5', 'ra0': 3.794, 'dec0': 32.865, 'vlsr': 9.0},
    'NGC1333': {'source': 'NGC1333*', 'source_out': 'NGC1333', 'ra0': 3.4862, 'dec0': 31.23, 'vlsr': 5.0},
    'PER2': {'source': 'PER2*', 'source_out': 'Per-emb-2', 'ra0': 3.53806, 'dec0': 30.830, 'vlsr': 5.0},
    'CLOUDH': {'source': 'CLOUDH', 'source_out': 'CloudH', 'ra0': 18.9521, 'dec0': 2.1775, 'vlsr': 45.0}
}

line_catalogue = 'line_catalogue_3mm.csv'
# Folder definitions
uvt_dir = 'D/'
uvt_dir_out = 'D30m/'
dir_30m = '30m/'
inputdir = 'raw_data/'

ignorefiles = ['raw_data/FTSOdp20220220.30m',
               'raw_data/FTSOdp20220731.30m']


# Define lines, frequencies as used with NOEMA data
# vel_ext is better to give a wide range
# vel_win is the velocity window for the baseline, better tight around the line
# obs_param = {
#     '13CS': {'freq': 92494.3080, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'CCS_93': {'freq': 93870.0980, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'CCD': {'freq': 72107.7205, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'c-C3H': {'freq': 91494.3490, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'HNCO_87_10K': {'freq': 87925.23700, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'HNCO_87_53K': {'freq': 87597.33000, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'DCOp': {'freq': 72039.31220, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'HCOp': {'freq': 89188.52470, 'vel_ext': '-85 95', 'vel_win': '-45 60'},
#     'H13COp': {'freq': 86754.22840, 'vel_ext': '-35 35', 'vel_win': '-5 20'},
#     'DCN': {'freq': 72414.90500, 'vel_ext': '-35 45', 'vel_win': '-3.5 1.5 5 10 12 16'},
#     'HCN': {'freq': 88631.8475, 'vel_ext': '-55 75', 'vel_win': '-23 30'},
#     'H13CN': {'freq': 86339.9214, 'vel_ext': '-35 45', 'vel_win': '-2 1 4 15'},
#     'HC15N': {'freq': 86054.9610, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     'H15NC': {'freq': 88865.692, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     'HN13C': {'freq': 870908500, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     'DNC': {'freq': 76305.72700, 'vel_ext': '-35 45', 'vel_win': '5 12'},
#     'HNC': {'freq': 90663.56800, 'vel_ext': '-35 45', 'vel_win': '2 15'},
#     'H2CO': {'freq': 72837.94800, 'vel_ext': '-35 45', 'vel_win': '-5 20'},
#     'HC3N_72': {'freq': 72783.82200, 'vel_ext': '-35 45', 'vel_win': '0 15'},
#     'HC3N_91': {'freq': 90979.02300, 'vel_ext': '-35 45', 'vel_win': '0 15'},
#     'N2Dp': {'freq': 77109.24330, 'vel_ext': '5 95', 'vel_win': '31 37 41 56'},
#     'N2Hp': {'freq': 93173.39770, 'vel_ext': '5 95', 'vel_win': '30 58'},
#     'SO2': {'freq': 72758.24340, 'vel_ext': '-35 45', 'vel_win': '0 15'},
#     'SiO': {'freq': 86846.96, 'vel_ext': '-70 95', 'vel_win': '-10 40'},
#     'SO': {'freq': 86093.95000, 'vel_ext': '-70 95', 'vel_win': '-10 40'},
#     'CCH_873_a': {'freq': 87284.105, 'vel_ext': '-35 45', 'vel_win': '2 15'},
#     'CCH_873_b': {'freq': 87316.925, 'vel_ext': '-25 45', 'vel_win': '5 12'},
#     'CCH_873_c': {'freq': 87328.624, 'vel_ext': '-35 45', 'vel_win': '2 15'},
#     'CH3CHO_93.5_a': {'freq': 93580.90910, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     'CH3CHO_93.5_b': {'freq': 93595.23490, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     # CH3OH
#     'CH3OH_76.5': {'freq': 76509.684, 'vel_ext': '-35 45', 'vel_win': '5 10'},
#     'CH3OH_92.4': {'freq': 92409.490, 'vel_ext': '-35 45', 'vel_win': '5 10'},
# }
